pennylane/devices/preprocess.py

```diff
@@ -545,3 +544,211 @@ def _validate_snapshot_shots(tape, sample_measurements, analytic_measurements, n
                     raise qml.DeviceError(
                         f"Measurement {m} not accepted for analytic simulation on {name}."
                     )
+
+@transform
+def measurements_from_samples(tape):
+    """Quantum function transform that replaces all measurements from a tape with a single
+    sample measurement, and adds postprocessing functions for each original measurement.
+
+    This transform can be used to make tapes compatible with device backends that only return
+    `samples`. The final output will return the initial requested measurements, calculated from
+    the raw counts returned immediately after execution.
+
+    The transform is only applied if the tape is being executed with shots.
+
+    .. note::
+        This transform diagonalizes all the operations on the tape. An error will
+        be raised if non-commuting terms are encountered. To avoid non-commuting
+        terms in circuit measurements, the :func:`split_non_commuting <pennylane.transforms.split_non_commuting>`
+        transform can be applied.
+
+    Args:
+        tape (QNode or QuantumTape or Callable): A quantum circuit.
+
+    Returns:
+        qnode (QNode) or quantum function (Callable) or tuple[List[QuantumTape], function]: The
+        transformed circuit as described in :func:`qml.transform <pennylane.transform>`.
+
+    **Example**
+
+    Consider the tape:
+
+    >>> ops = [qml.X(0), qml.RY(1.23, 1)]
+    >>> measurements = [qml.expval(qml.Y(0)), qml.probs(wires=[1])]
+    >>> tape = qml.tape.QuantumScript(ops, measurements, shots=10)
+
+    We can apply the transform to diagonalize and convert the two measurements to a single sample:
+
+    >>> (new_tape, ), fn = qml.transforms.measurements_from_samples(tape)
+    >>> new_tape.measurements
+    [sample(wires=[0, 1])]
+
+    The tape operations now include diagonalizing gates.
+
+    >>> new_tape.operations
+    [X(0), RY(1.23, wires=[1]), RX(1.5707963267948966, wires=[0])]
+
+    Executing the tape returns samples that can be post-processed to get the originally requested measurements:
+
+    >>> dev = qml.device("default.qubit")
+    >>> dev.execute(new_tape)
+    array([[1, 0],
+           [0, 0],
+           [0, 1],
+           [1, 1],
+           [0, 1],
+           [1, 0],
+           [0, 1],
+           [1, 0],
+           [1, 0],
+           [1, 0]])
+
+    >>> fn((res,))
+    (-0.2, array([0.6, 0.4]))
+    """
+    if tape.shots.total_shots is None:
```

Comment:
```suggestion
    if not tape.shots:
```
Purely stylistic, but we can take advantage of the booleaness of the `Shots` class.